Use Webview to manage cluster settings

angelozerr opened the pull request:
Use Webview to manage cluster settings

Fixes #88

Signed-off-by: azerr <azerr@redhat.com>

fbricon commented:
<img width="1189" alt="Screenshot 2021-05-26 at 12 05 54" src="https://user-images.githubusercontent.com/148698/119642210-c3e62880-be1a-11eb-9c1a-8b3c76e1da8a.png">
- Page title should be "Edit Kafka cluster" or (New Kafka cluster)
- Header on top should be removed
- Cluster block should move up
- labels and text fields should be aligned and having the same width
- Authentication "None" should have a label
- Hide Back/Next, Finish -> Save

angelozerr commented:
Please note now I'm consuming 0.2.5 vscode-wizard which has been published (which fixes some bugs that you mentionned)

> Page title should be "Edit Kafka cluster" or (New Kafka cluster)

Fixed (for the moment I manage only Edit)

> Header on top should be removed

Fixed by 0.2.5

> Cluster block should move up

Sorry I don't understand what you mean?

> labels and text fields should be aligned and having the same width

Fixed, but there is a problem with select that I have reported in https://github.com/redhat-developer/vscode-wizard/issues/33

> Authentication "None" should have a label

Fixed

> Hide Back/Next, Finish -> Save

TODO

angelozerr commented:
I pushed a new commit, I have started to declare the UI for SSL:

![image](https://user-images.githubusercontent.com/1932211/121157813-27744b00-c84a-11eb-9642-560be9563e9a.png)

Please note:

 * we should manage a browse button to select the proper files. See https://github.com/redhat-developer/vscode-wizard/issues/34
 * for the moment it's the UI, I need to write code to use values of this fields.

rgrunber commented:
I'd probably use the full name of the term.
Ca -> Certificate Authority (CA)
Cert -> Certificate

I don't think the SSL checkbox should exist. The underlying field, `cluster.ssl` should simply be set to true when the other 3 fields are valid/non-empty.

fbricon commented:
@rgrunber no you can set SSL=true without specifying certs

angelozerr commented:
>I'd probably use the full name of the term.
>Ca -> Certificate Authority (CA)
>Cert -> Certificate

Fixed

>I don't think the SSL checkbox should exist. The underlying field, cluster.ssl should simply be set to true when the other 3 fields are valid/non-empty.

As @fbricon said we can enable SSL without certificates

angelozerr commented:
Just for your information I created a PR https://github.com/redhat-developer/vscode-wizard/pull/37 to improve UI, here the result:

![image](https://user-images.githubusercontent.com/1932211/121357206-e192b200-c931-11eb-9f28-28107a039a22.png)

angelozerr commented:
The PR is ready now. The only thing that I should add is the passphrase field (I think), but I'm wiating for feedback from https://github.com/jlandersen/vscode-kafka/issues/191#issuecomment-871999125

fbricon commented:
need to document that you need to run "npm run compile" at least once before debugging the extension in vscode

fbricon commented:
<img width="1155" alt="Screenshot 2021-07-01 at 15 37 34" src="https://user-images.githubusercontent.com/148698/124133621-4b5e3180-da82-11eb-925f-ca01d829e7f9.png">

- New Cluster(s) title should be "Add New Cluster(s)"
- cluster provider: remove the "none" option, have "configure manually"  preselected
- cluster provider combo should not crop selected provider

If vscode-rhoas is installed, you select the Red Hat provider, click finish, the wizard stays opened. It should close

fbricon commented:
More to fix:

<img width="1216" alt="Screenshot 2021-07-01 at 15 46 08" src="https://user-images.githubusercontent.com/148698/124136301-f2dc6380-da84-11eb-8905-081c3f58197c.png">

fbricon commented:
Input text alignment is borked in case of an error
<img width="867" alt="Screenshot 2021-07-01 at 15 56 14" src="https://user-images.githubusercontent.com/148698/124136704-4f3f8300-da85-11eb-8096-3c8ff52e1b4d.png">

fbricon commented:
When selecting SASL/*, the SSL box is not automatically checked, rendering the cluster unable to connect.

fbricon commented:
When creating a new cluster, the wizard is stil opened after saving. If you click save again, a "Cluster name 'foo' already exists." error is shown. The wizard should be closed, or at the very least should enter edit mode.

fbricon commented:
When editing a new cluster:
- change name
- save
- check ssl
- save => cluster name already exists

fbricon commented:
In the SSL section, rename:
Key => Client key
Certificate => Client certificate

fbricon commented:
certificate/key files should allow ".pem" and other file extensions (see https://kafka.js.org/docs/configuration#ssl and https://www.ssl.com/guide/pem-der-crt-and-cer-x-509-encodings-and-conversions/)

angelozerr commented:
> New Cluster(s) title should be "Add New Cluster(s)"

fixed

> cluster provider: remove the "none" option, have "configure manually" preselected

fixed

> cluster provider combo should not crop selected provider

it's an issue for vscode-wizard, see https://github.com/redhat-developer/vscode-wizard/issues/58

> If vscode-rhoas is installed, you select the Red Hat provider, click finish, the wizard stays opened. It should close. If you click finish a 2nd time, it closes.

That's strange, I cannot reproduce on Windows OS?

fbricon commented:
When renaming a cluster, the selected cluster in the status bar is not renamed

angelozerr commented:
> When selecting SASL/*, the SSL box is not automatically checked, rendering the cluster unable to connect.

I cannot do that with vscode-wizard, but I validate that:

![image](https://user-images.githubusercontent.com/1932211/124252860-b7df3c00-db27-11eb-8442-8ab803ac2fc7.png)

> When creating a new cluster, the wizard is stil opened after saving

I cannot reproduce it -(

> If you click save again, after checking the ssl box, a "Cluster name 'foo' already exists." error is shown. The wizard should be closed, or at the very least should enter edit mode.

fixed

> When editing an existing cluster:
>
>change name
>save
>check SSL => cluster name already exists

Same error than below, fixed

> certificate/key files should allow ".pem" and other file extensions (see https://kafka.js.org/docs/configuration#ssl and https://www.ssl.com/guide/pem-der-crt-and-cer-x-509-encodings-and-conversions/)

I add other extensions according to https://www.tutorialsteacher.com/https/ssl-certificate-format 

Here the filters now:

![image](https://user-images.githubusercontent.com/1932211/124253468-579cca00-db28-11eb-89be-734529e49918.png)

I allow *.* because I fear to forget some file extensions.

angelozerr commented:
> When renaming a cluster, the selected cluster in the status bar is not renamed

fixed

angelozerr commented:
@fbricon  everything should be fixed exept:

 * UI problem, I have created an issue in vscode-wizard https://github.com/redhat-developer/vscode-wizard/issues/58
 * disabled of  fileds according to combo mechanism, select SSL checkbox when mechanims != auto, I hav ecreated an issue in vscode-wizard https://github.com/redhat-developer/vscode-wizard/issues/59

fbricon commented on src/wizards/validators.ts:
```diff
@@ -1,27 +1,33 @@
 }
 
+export function validateFile(filePath: string) : string | undefined {
+    if (!fs.existsSync(filePath)) {
```
Should check it's an actual file

angelozerr replied:
fixed

fbricon commented on src/wizards/validators.ts:
```diff
@@ -1,27 +1,36 @@
+        return `The file '${filePath}' doesn't exists.`;
+    }
+    if (fs.lstatSync(filePath).isDirectory()) {
+        return `The '${filePath}' is a directory, please select a file.`;
```
'${filePath}' is a directory, please select a file.

angelozerr replied:
fixed

fbricon commented on src/wizards/validators.ts:
```diff
@@ -1,27 +1,36 @@
 
+export function validateFile(filePath: string) : string | undefined {
+    if (!fs.existsSync(filePath)) {
+        return `The file '${filePath}' doesn't exists.`;
```
doesn't exist.

angelozerr replied:
fixed

fbricon commented:
Nice work @angelozerr! Thanks!
